rfd/0053-passwordless-fido2.md

```diff
@@ -0,0 +1,494 @@
+https://github.com/keys-pub/go-libfido2/blob/9246c0bee502c5f2c2495557f07fa5f45da8cd1c/fido2.go#L770)
+as the method to choose the authenticator. This technique requires one user
+touch/verification and avoids the need for PINs with biometric authenticators.
+
+(3) is when we enumerate resident keys for the {authenticator, RPID} pair. We
+can skip user interaction if only a single key exists.
+
+(4) is the assertion step, where the authenticator signs the challenge. This
+step requires a final touch/verification from the user.
+
+\* currently go-libfido2 doesn't expose operations equivalent to
+[fido_dev_touch_begin](https://developers.yubico.com/libfido2/Manuals/fido_dev_get_touch_begin.html)/[
+fido_dev_touch_status](https://developers.yubico.com/libfido2/Manuals/fido_dev_get_touch_status.html),
+but we can make do by attempting an [Assertion](
+https://github.com/keys-pub/go-libfido2/blob/9246c0bee502c5f2c2495557f07fa5f45da8cd1c/fido2.go#L639).
+(We could fork it and add it ourselves; I have a local patch that does just
+that, but I'd rather keep light on forking if possible.)
+
+#### libfido2 and Teleport
+
+[libfido2](https://github.com/Yubico/libfido2) adds a native dependency to
+Teleport. Its Go wrapper,
+[go-libfido2](https://github.com/keys-pub/go-libfido2/), does little to
+alleviate the pains of dealing with a native dependency (in fact, as a wrapper,
+there is little it could do).
+
+libfido2 itself depends on [libcbor](https://github.com/pjk/libcbor), OpenSSL
+1.1 or newer, [zlib](https://zlib.net/) and libudev (Linux only, part of
+systemd).
+
+Care must be taken to produce deterministic (and equivalent!) builds for the
+various supported operating systems - as a last resort, building libcbor and
+libfido2 from source may be necessary (pulling the more complex, less sensitive
+libraries from package managers).
+
+Linking for go-libfido2 varies depending on OS - it defaults to
+[static for macOS](
+https://github.com/keys-pub/go-libfido2/blob/master/fido2_static.go) (with an
+embedded libcbor binary) and [dynamic for Linux and Windows](
+https://github.com/keys-pub/go-libfido2/blob/master/fido2_other.go#L6) (with
+various embedded binaries for Windows). We should consider replacing those
+during build, or once again forking go-libfido2 in order to standardize building
+and linking strategies. (Static linking seems to be the best option for `tsh`.)
+
+During development, the libfido2-aware code in Teleport is to be protected by a
+`libfido2` build tag - said tag could be kept in the final product if dealing
+with the library proves to be a hassle.
+
+### Security
+
+Security considerations are largely unchanged in relation to the
+[Passwordless RFD][passwordless rfd].
+
+It's important to note that CLI applications with hardware key access are
+particularly dangerous, as they are not subject to the security measures
+we've come to expect from browsers (in fact, _any_ untrusted application that
```

Comment:
Would `tsh` need any special permissions to access hardware keys on any of the supported OS? If any, can they be requested at runtime? I'm hoping we don't need to run as root on linux or with admin rights on Windows, or something like that.

ninja edit - I see this is covered in the next doc for MacOS. Do we need to do anything similar for Linux or Windows?

Reply:
Nope, we don't need special permissions for FIDO2 - in fact it's the exact same thing we do today for U2F. That is why you should be wary of apps interacting with your Yubikeys - they could be pretending to be anyone.

MacOS / Touch ID is a different beast, in general it's much safer in terms of permissions and app isolation, but one should still be careful with whom they trust.